### DISCUSSION TITLE
ExportDXF:  arcs come out as polylines

### DISCUSSION THREAD

**Discussion by shrekuntu** [gumyr/build123d#773] [General]
I'm finding that when entities (fillets, slots etc) are projected to dxf and should be arcs, in dxf they come out as some polylines that can't be dimensioned as radii.
Is there any workaround for this?

**Comment by gumyr** [comment]
This works for me:
```py
from build123d import *
from ocp_vscode import show_all
import ezdxf

a = CenterArc((0, 0), 1, 0, 90)

exporter = ExportDXF(unit=Unit.MM, line_weight=0.5)
exporter.add_layer("Layer 1")
exporter.add_shape(a, layer="Layer 1")
exporter.write("output.dxf")


doc = ezdxf.readfile("output.dxf")
# Iterate over all entities in the modelspace and print their types
for entity in doc.modelspace():
    if entity.dxftype() == "ARC":
        # Display ARC-specific attributes
        print("ARC found:")
        print(f"  Center: {entity.dxf.center}")
        print(f"  Radius: {entity.dxf.radius}")
        print(f"  Start Angle: {entity.dxf.start_angle}")
        print(f"  End Angle: {entity.dxf.end_angle}")
    else:
        print(entity.dxftype())
        
show_all()
```
```
ARC found:
  Center: (0.0, 0.0, 0.0)
  Radius: 1.0
  Start Angle: 0.0
  End Angle: 90.0
```

How were the DXF files generated?

**Reply by shrekuntu** [reply]
This is how I generated a projection and saved it to dxf:
I have a shaft turndown (fillet) and keyway (slot) which come out as some polyline in qcad where I inspect it.
Other entities look good (straight lines).

```
    view_port_origin=(0, 0, 90)
    visible, hidden = shaft.part.project_to_viewport(view_port_origin)
    max_dimension = max(*Compound(children=visible + hidden).bounding_box().size)    
    dxf = ExportDXF(unit=Unit.IN, line_weight=0.5)
    dxf.add_layer("ent", line_type=LineType.CONTINUOUS)
    dxf.add_shape(visible, layer="ent")
    dxf.write("shaft_projection.dxf")
```

**Reply by gumyr** [reply]
Did you look at the `visible` edges - are they arcs or lines?

**Reply by shrekuntu** [reply]
This is what I get when I do:
```
for v in visible:
    print(v.geom_type)
```

```
GeomType.BSPLINE
GeomType.ELLIPSE
GeomType.LINE
GeomType.BSPLINE
GeomType.BSPLINE
GeomType.LINE
GeomType.BSPLINE
GeomType.LINE
GeomType.LINE
GeomType.BSPLINE
GeomType.LINE
GeomType.BSPLINE
GeomType.LINE
GeomType.BSPLINE
GeomType.ELLIPSE
GeomType.LINE
GeomType.BSPLINE
GeomType.BSPLINE
GeomType.BSPLINE
GeomType.BSPLINE
GeomType.LINE
GeomType.LINE
GeomType.BSPLINE
GeomType.LINE
```

**Reply by gumyr** [reply]
The projection might have converted some circles to bsplines thus loosing the ability to use the `radius` property but given the small number of lines it didn't approximate the curves with many straight line segments.  Viewers typically do this though - could it be that the curves are smooth but the viewer is rendering them as a series of straight lines?

It is possible to create a `section` instead of projection? The circles are more likely to remain circles.

**Reply by shrekuntu** [reply]
Here is a full example with a fillet:
It shows ellipses and bsplines. visible loop shows only 4 lines, while I think it should be 8. So some lines are approximated by other types?

Maybe its a cylindrical surface on which these fillets/slots are made. On your example ex35 slots, visible edges are arcs but it's a flat  surface

Not sure if section would help, I may loose clean view of the keyways, while possibly solving shaft turndown fillet?

At this point it's not a critical to have these as arcs, I don't have to dimension them, worst case I can draw a leader dim.
Maybe down the road will be a better solution.

Thank you for your help.

```
from build123d import *

with BuildPart() as shaft:
    with BuildSketch(Plane.YZ) as s1:
        Circle(1.75)
    extrude(amount=5)
    
    with BuildSketch(shaft.faces().sort_by(Axis.X)[-1]):
        Circle(1.5)
    extrude(amount=2)
    
    fillet(shaft.edges().filter_by(GeomType.CIRCLE)
        .sort_by_distance((5, 0, 0))[0], 0.1)
    
show_object(shaft)

view_port_origin=(0, 0, 90)
visible, hidden = shaft.part.project_to_viewport(view_port_origin)
max_dimension = max(*Compound(children=visible + hidden).bounding_box().size)    
dxf = ExportDXF(unit=Unit.IN, line_weight=0.5)
dxf.add_layer("ent", line_type=LineType.CONTINUOUS)
dxf.add_shape(visible, layer="ent")
dxf.write("shaft_simple.dxf")
```
and visible edges are:
```
GeomType.ELLIPSE
GeomType.LINE
GeomType.ELLIPSE
GeomType.ELLIPSE
GeomType.LINE
GeomType.ELLIPSE
GeomType.ELLIPSE
GeomType.BSPLINE
GeomType.BSPLINE
GeomType.LINE
GeomType.LINE
```

**Reply by gumyr** [reply]
It's working as far as I can tell.  Here I extend your example by recovering build123d object from the DXF objects and they match perfectly:
```py
with BuildPart() as shaft:
    with BuildSketch(Plane.YZ) as s1:
        Circle(1.75)
    extrude(amount=5)

    with BuildSketch(shaft.faces().sort_by(Axis.X)[-1]):
        Circle(1.5)
    extrude(amount=2)

    fillet(shaft.edges().filter_by(GeomType.CIRCLE).sort_by_distance((5, 0, 0))[0], 0.1)

view_port_origin = (0, 0, 90)
visible, hidden = shaft.part.project_to_viewport(view_port_origin)
visible_ellipse = visible.filter_by(GeomType.ELLIPSE)
visible_bspline = visible.filter_by(GeomType.BSPLINE)
visible_line = visible.filter_by(GeomType.LINE)
max_dimension = max(*Compound(children=visible + hidden).bounding_box().size)
dxf = ExportDXF(unit=Unit.IN, line_weight=0.5)
dxf.add_layer("ent", line_type=LineType.CONTINUOUS)
dxf.add_shape(visible, layer="ent")
dxf.write("shaft_simple.dxf")

doc = ezdxf.readfile("shaft_simple.dxf")
recovered = ShapeList()
for entity in doc.modelspace():
    print(entity.dxftype())
    if entity.dxftype() == "LINE":
        recovered.append(Line(entity.dxf.start, entity.dxf.end))
    elif entity.dxftype() == "ELLIPSE":
        center = entity.dxf.center
        major_axis = entity.dxf.major_axis
        x_radius = (major_axis[0] ** 2 + major_axis[1] ** 2) ** 0.5
        y_radius = x_radius * entity.dxf.ratio
        rotation = math.degrees(math.atan2(major_axis[1], major_axis[0]))
        start_angle = math.degrees(entity.dxf.start_param)
        end_angle = math.degrees(entity.dxf.end_param)
        recovered.append(
            EllipticalCenterArc(
                center=center,
                x_radius=x_radius,
                y_radius=y_radius,
                start_angle=start_angle,
                end_angle=end_angle,
                rotation=rotation,
            )
        )
    elif entity.dxftype() == "SPLINE":
        control_points = [(point[0], point[1]) for point in entity.control_points]
        start_tangent = entity.dxf.get("start_tangent")
        end_tangent = entity.dxf.get("end_tangent")
        if any(t is None for t in [start_tangent, end_tangent]):
            tangents = ()
        else:
            tangents = (start_tangent, end_tangent)
        recovered.append(Spline(*control_points, tangents=tangents))

show(shaft, visible, visible_ellipse, visible_bspline, visible_line, recovered)
```
![image](https://github.com/user-attachments/assets/67076660-4bb0-42a3-a1c0-19ad6b558b82)

**Comment by gumyr** [comment]
Did you look at using `section` instead of a projection?

If you do:
```py
visible = section(shaft.part, Plane.XY).edges()
```
you'll get:
![image](https://github.com/user-attachments/assets/38a30764-d47e-4e85-bac9-cb4113455781)
which results in only `LINE` and `ARC` objects in the DXF.

**Reply by shrekuntu** [reply]
Thank you very much @gumyr, you went to great extent to help me out, it's greatly appreciated.
I haven't tried section as I wasn't sure how to do it, now I see it, but in my industry it's not common to cross section/hatch shafts
Your other solution would work better but still has an curved line for end of the shafts, it should be a straight line.
I've attached a screenshot of how it should look. I'm OK with this for now, best compromise at the moment.
The only thing I am concerned how will I select leftmost and rightmost point on the shaft for making shaft length dimension via ezdxf due to this distortion but it looks like a straight line to me?
![shaft](https://github.com/user-attachments/assets/b8dca9d0-40c0-4993-82bd-79dd9f6d2db1)

**Reply by gumyr** [reply]
Are you trying to create a dimensioned drawing of your part?  If so you probably want to start with a projection that's "straight down" like this:
```py
visible, hidden = shaft.part.project_to_viewport(
    viewport_origin=(0, 0, 10), viewport_up=(0, 1, 0), look_at=(0, 0, 0)
)
```
![image](https://github.com/user-attachments/assets/a48c98b5-7ca2-4481-853e-cd66dcf9d016)
(I added a slot)

By default the `project_to_viewport` method will `look_at` the center of the part which is difficult to predict with a complex part like this.  The `viewport_up` controls the rotation of the page so you could orient it vertically if you like.

Unfortunately, this projection will still convert straight lines and one of the arcs to bsplines so it isn't the DXF export that causing the problem it's the projection itself and that can't be fixed (without a great deal of effort) as this functionality is provided by the OCCT CAD kernel not build123d directly.

**Reply by shrekuntu** [reply]
Sample you posted looks straight lines, I don't see any distortion.. I will try it. I may not need to query ezdxf points much, as I already have my main parameters, so I can generate points for dimensions. 

Meantime, I had an idea based on ex35 (slots) on a flat plate, which gets lines and arcs. So in parallel with a round shaft, I make a "square" shaft which has perfect projection, I already tested it. Then I can assemble top view from square shaft and side and iso views from the real shaft. Nobody could ever notice that it comes from a different model.

Thank you.
